activation/activation_test.go

```diff
@@ -271,48 +159,87 @@ func assertLastAtx(r *require.Assertions, posAtx, prevAtx *types.VerifiedActivat
 	r.Equal(poetRef, atx.GetPoetProofRef())
 }
 
-func publishAtx(t *testing.T, b *Builder, clockEpoch types.EpochID, buildNIPostLayerDuration uint32) (published, builtNIPost bool, err error) {
+func publishAtx(
+	t *testing.T,
+	tab *testAtxBuilder,
+	posAtxId types.ATXID,
+	posAtxLayer types.LayerID,
+	currLayer *types.LayerID, // pointer to keep current layer consistent across calls
+	buildNIPostLayerDuration uint32,
+	publishErr error,
+) (*types.ActivationTx, error) {
 	t.Helper()
-	net.lastTransmission = nil
-
-	nipostBuilderMock.mu.Lock()
-	nipostBuilderMock.buildNIPostFunc = func(challenge *types.PoetChallenge) (*types.NIPost, time.Duration, error) {
-		builtNIPost = true
-		layerClockMock.mu.Lock()
-		layerClockMock.currentLayer = layerClockMock.currentLayer.Add(buildNIPostLayerDuration)
-		layerClockMock.mu.Unlock()
-		hash, err := challenge.Hash()
-		require.NoError(t, err)
-		return newNIPostWithChallenge(hash, poetBytes), 0, nil
-	}
-	nipostBuilderMock.mu.Unlock()
-
-	layerClockMock.mu.Lock()
-	layerClockMock.currentLayer = clockEpoch.FirstLayer().Add(3)
-	layerClockMock.mu.Unlock()
-
-	err = b.PublishActivationTx(context.Background())
 
-	nipostBuilderMock.mu.Lock()
-	nipostBuilderMock.buildNIPostFunc = nil
-	nipostBuilderMock.mu.Unlock()
-
-	return net.lastTransmission != nil, builtNIPost, err
+	publishEpoch := posAtxLayer.GetEpoch() + 1
+	tab.mclock.EXPECT().GetCurrentLayer().DoAndReturn(
+		func() types.LayerID {
+			return *currLayer
+		}).AnyTimes()
+	tab.mhdlr.EXPECT().GetPosAtxID().Return(posAtxId, nil)
+	tab.mclock.EXPECT().LayerToTime(gomock.Any()).DoAndReturn(
+		func(got types.LayerID) time.Time {
+			require.Equal(t, publishEpoch.FirstLayer(), got)
+			return time.Now().Add(100 * time.Second)
+		})
+	lastOpts := DefaultPostSetupOpts()
+	tab.mpost.EXPECT().LastOpts().Return(&lastOpts).AnyTimes()
+	tab.mnipost.EXPECT().BuildNIPost(gomock.Any(), gomock.Any(), gomock.Any()).DoAndReturn(
+		func(_ context.Context, challenge *types.PoetChallenge, _ time.Time) (*types.NIPost, time.Duration, error) {
+			*currLayer = currLayer.Add(buildNIPostLayerDuration)
+			return newNIPostWithChallenge(challenge.Hash(), []byte("66666")), 0, nil
+		})
+	ch := make(chan struct{}, 1)
+	close(ch)
+	tab.mclock.EXPECT().AwaitLayer(publishEpoch.FirstLayer()).DoAndReturn(
+		func(got types.LayerID) chan struct{} {
+			// advance to publish layer
+			if currLayer.Before(got) {
+				*currLayer = got
+			}
+			return ch
+		})
```

Comment:
The current layer shared like that has been a source of data races in the past because the mocks and tests have been accessing it concurrently). Is it safe now?

Reply:
i think so. that's why i used a `*types.LayerID` here so the mock and test is referencing the same object.